src/agent.ts

```diff
@@ -84,43 +84,37 @@ export class Agent {
 
     return agent;
   }
-
+  
   public async run(): Promise<void> {
     try {
       while (true) {
-        if (!this._executedLastIteration) {
-          // Ask the user for input
-          await this._askUserForPrompt();
-        }
-
-        // Get a response from the AI
-        const response = await this._askAiForResponse();
-
-        // Process response, and extract function call
-        const functionCall = this._processAiResponse(response);
-
-        if (functionCall) {
-          // Get confirmation from the user
-          const confirmation = await this._askUserForConfirmation(
-            functionCall
-          );
-
-          if (confirmation) {
-            // Execute function calls
-            await this._executeFunctionCall(functionCall);
-          } else {
-            // Execute a NOOP
-            this._executeNoop(functionCall);
-            this._executedLastIteration = false;
+        let response: OpenAIResponse | null = null;
+        let functionCall: OpenAIFunctionCall | undefined = undefined;
+        do {
+          if (!this._executedLastIteration) {
+            // Ask the user for input
+            await this._askUserForPrompt();
+          }
+  
+          // Get a response from the AI
+          response = await this._askAiForResponse();
+  
+          // Process response, and extract function call
+          functionCall = this._processAiResponse(response);
+  
+          if (functionCall) {
+            // Check if the function exists, prompt for approval, and execute
+            await this._executeFunctionIfExists(functionCall);
           }
-        } else {
-          this._executedLastIteration = false;
-        }
+        } while (this._executedLastIteration && functionCall);
```

Comment:
How does the flow go with this condition and the outer loop? If there is indeed a function call and it gets executed this would iterate; but if that condition isn't met, then the outer while(true) loop would execute which would make the inner loop execute again.